src/veh_interact.cpp

```diff
@@ -572,13 +584,46 @@ void veh_interact::do_install()
     }
     mvwprintz(w_mode, 0, 1, c_ltgray, _("Choose new part to install here:"));
     wrefresh (w_mode);
-    
-    int pos = 0;
+
+    std::array<std::string,7> tab_list = {_("All"),_("Cargo"),_("Light"),_("Util"),_("Hull"),_("Armor"),_("Other")};
+    std::array <std::function<bool(vpart_info)>,7> tab_filters; // filter for each tab, last one
+    tab_filters[0] = [&](vpart_info part) { return true; }; // All
+    tab_filters[1] = [&](vpart_info part) { return part.has_flag(VPFLAG_CARGO); }; // Cargo
+    tab_filters[2] = [&](vpart_info part) { return part.has_flag(VPFLAG_LIGHT) || part.has_flag(VPFLAG_CONE_LIGHT) ||  // Light
+                                                   part.has_flag(VPFLAG_CIRCLE_LIGHT) || part.has_flag(VPFLAG_DOME_LIGHT) ||
+                                                   part.has_flag(VPFLAG_AISLE_LIGHT) || part.has_flag(VPFLAG_EVENTURN) ||
+                                                   part.has_flag(VPFLAG_ODDTURN); };
+    tab_filters[3] = [&](vpart_info part) { return part.has_flag(VPFLAG_TRACK) || part.has_flag(VPFLAG_FRIDGE) || //Util
+                                                   part.has_flag("KITCHEN") || part.has_flag("WELDRIG") || part.has_flag("CRAFTRIG") ||
+                                                   part.has_flag("CHEMLAB") || part.has_flag("FORGE") || part.has_flag("HORN") ||
+                                                   part.has_flag(VPFLAG_RECHARGE) || part.has_flag(VPFLAG_EXTENDS_VISION) ||
+                                                   part.has_flag("POWER_TRANSFER"); };
+    tab_filters[4] = [&](vpart_info part) { return part.has_flag(VPFLAG_OBSTACLE); }; // Hull
+    tab_filters[5] = [&](vpart_info part) { return part.has_flag(VPFLAG_ARMOR); }; // Armor
+    tab_filters[tab_filters.size()-1] = [&](vpart_info part) { // Other: everything that's not in the other filters
+        for (int i=1; i < tab_filters.size()-1; i++ ) {
+            if( tab_filters[i](part) ) return false;
+        }
+        return true; };
```

Comment:
Oh, hello, Mr. `O(N**2)`. Good thing `N` is small. How often are we calling this?